File: pyround1/1154.一年中的第几天.py
#
# @lc app=leetcode.cn id=1154 lang=python3
#
# [1154] 一年中的第几天
#

# @lc code=start
class Solution:
    def dayOfYear(self, date: str) -> int:
        year = int(date[0:4])
        month = int(date[5:7])
        day = int(date[8:])
        dic = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
        if year % 400 == 0 or year% 4 == 0 and year % 100 != 0:
            dic[1] = 29
        return sum(dic[:month-1]) + day
        
# @lc code=end

File: 1154.一年中的第几天.cpp
/*
 * @lc app=leetcode.cn id=1154 lang=cpp
 *
 * [1154] 一年中的第几天
 */

// @lc code=start
class Solution {
public:
    int days[2][13] = {{0, 31,28, 31, 30, 31, 30, 31,31,30,31, 30, 31},
                        {0, 31,29, 31, 30, 31, 30, 31,31,30,31, 30, 31}};

    bool leap(int y)
    {
        return (!(y % 4) && (y % 100)) || !(y % 400);
    }
    int dayOfYear(string date) {
        int y, m, d, cnt = 1, dc = 1, mc = 1;
        sscanf(date.c_str(), "%d-%d-%d", &y, &m, &d);

        /*从1月1号算起*/
        while(dc < d || mc < m){
            dc++;
            if(dc == days[leap(y)][mc] + 1){  //满当月天数
                mc++;
                dc = 1;          //日期变为下一月的1号
            }
            cnt++;
        }
        return cnt;

    }
};
// @lc code=end

